src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,317 @@
+        //
+        // ⚠ DO NOT AWAIT INSIDE THE USING. The Dispose method that restores ExecutionContext flow must run on the
+        // same thread where SuppressFlow was originally run.
+        using var _ = FlowControlHelper.TrySuppressFlow();
+        return ReadDataSuppressedFlowAsync(pipeReader, objectCount, serializerService, callback, arg, cancellationToken);
 
-            for (int i = 0, n = objectCount; i < n; i++)
+        static async ValueTask ReadDataSuppressedFlowAsync(
+            PipeReader pipeReader, int objectCount, ISerializerService serializerService, Action<Checksum, T, TArg> callback, TArg arg, CancellationToken cancellationToken)
+        {
+            using var pipeReaderStream = pipeReader.AsStream(leaveOpen: true);
+
+            for (var i = 0; i < objectCount; i++)
             {
+                // First, read the sentinel byte and the length of the data chunk we'll be reading.
+                const int HeaderSize = sizeof(byte) + sizeof(int);
+                var lengthReadResult = await pipeReader.ReadAtLeastAsync(HeaderSize, cancellationToken).ConfigureAwait(false);
+                if (lengthReadResult.IsCanceled || lengthReadResult.IsCompleted)
+                    return;
+
+                var (sentinelByte, length) = ReadSentinelAndLength(lengthReadResult);
+
+                // Check that the sentinel is correct.
+                Contract.ThrowIfTrue(sentinelByte != MessageSentinelByte);
+
+                // If so, move the pipe reader forward to the end of the header.
+                pipeReader.AdvanceTo(lengthReadResult.Buffer.GetPosition(HeaderSize));
+
+                // Now buffer in the rest of the data we need to read.  Because we're reading as much data in as
+                // we'll need to consume, all further reading (for this single item) can handle synchronously
+                // without worrying about this blocking the reading thread on cross-process pipe io.
+                var fillReadResult = await pipeReader.ReadAtLeastAsync(length, cancellationToken).ConfigureAwait(false);
+                if (fillReadResult.IsCanceled || fillReadResult.IsCompleted)
+                    return;
+
+                // Note: we have let the pipe reader know that we're done with 'read at least' call, but that we
+                // haven't consumed anything from it yet.  Otherwise it will throw that another read can't start
+                // from within ObjectReader.GetReader below.
+                pipeReader.AdvanceTo(fillReadResult.Buffer.Start);
+
+                // Now do the actual read of the data, synchronously, from the buffers that are now in memory within
+                // our process.  These reads will move the pipe-reader forward, without causing any blocking on
+                // async-io.
+                using var reader = ObjectReader.GetReader(pipeReaderStream, leaveOpen: true, cancellationToken);
```

Comment:
>ObjectReader.GetReader [](http://example.com/codeflow?start=35&length=22)

dumb question of the hour:

This can't just use a single ObjectReader instead of creating one every loop iteration?

Reply:
they are stateful.

Reply:
ok.  moved to single object reader.